### ISSUE TITLE
Killer Instinct Gold Skies Aren't Displaying Correctly

### ISSUE THREAD

**Issue by shinra358** [gonetz/GLideN64#406]
In Killer Instinct Gold, most of the skies are flickering, resetting in animations, or just not displaying correctly.

**Comment by AmbientMalice** [comment]
@shinra358 Could you give more details? They seem to be working for me.

**Comment by shinra358** [comment]
No they aren't. Take the time to look at the skies in attract mode and in most of the stages. I don't know how to further explain as I already explained everything. If they seem to be working fine for you, show us a video and I'll be happy to point it out.

**Comment by AmbientMalice** [comment]
By details, I meant what emulator you're using, what build, and perhaps an image of a corrupted sky.

I'll do some more testing later. I only briefly tested the other day on PJ64's latest source using HLE mode, and it seemed fine. (Using GLN64 Early Release build.) The problem with KI is that PJ64 has a recompiler issue that makes the health bars glitchy. Also, 32 bit mode was enabled in some older builds, which caused issues.

**Comment by shinra358** [comment]
PJ64 latest dev build (I think the same issue is on mupen as well). Latest glide build that a $40 contributor can get. Words describe it well without pics. PJ64 says problems rendering sky beside the game entry. So I believe it's been like that for a while.

**Comment by gonetz** [comment]
Could you check it with z64 plugin? I suppose that the problem is caused by LLE triangle command, which I ported from z64.

**Comment by AmbientMalice** [comment]
Seems to happen with Ziggy's LLE also, @gonetz

**Comment by gonetz** [comment]
Thanks!
I need help from RDP expert here.

**Comment by purplemarshmallow** [comment]
#459 could be related
I think water in goldeneye in Frigate level uses LLE triange command

**Comment by orbea** [comment]
This does not happen with mupen64plus + glide64 or glide64mk2 while its worse with z64.

This does happen with mupen64plus-libretro + glide64, but not in angrylion while gln64 and rice do not have it, but have another issue in its place.

This also does not happen on real hardware as seen in the below video.

https://youtube.com/watch?v=0-7sNrbvstk

apitrace - http://ks392457.kimsufi.com/orbea/stuff/trace/mupen64plus/gliden64_Killer_Instinct_Gold.trace.xz
apitrace log - http://pastebin.com/4weY4LMC

**Comment by fzurita** [comment]
This is still an issue. Someone online was kind enough to make a video of it:

https://www.youtube.com/watch?v=fT2AobCFDKE

**Comment by gonetz** [comment]
This issue will not gone until somebody will correct all issues with LLE Triangle.

**Comment by fzurita** [comment]
Is there a working implementation of LLE triangle that can be borrowed from?

**Comment by gonetz** [comment]
Sources of angrylion's plugin contain most complete and perfectly working implementation of LLE triangle. The question is how to rewrite it for hardware rendering.

**Comment by olivieryuyu** [comment]
that was already done actually but in Vulkan: https://github.com/libretro/mupen64plus-libretro/tree/master/mupen64plus-rsp-paraLLEl

**Comment by gonetz** [comment]
```
Code is heavily inspired by the Angrylion LLE plugin.
This plugin aims to be very similar to Angrylion LLE, but it is completely rewritten to run in parallel on GPUs.
```

I'm not sure that it is what we need. Careful reading needed though. And/or help from the author.
Also, I will not take that task, at least not soon.

**Comment by fzurita** [comment]
It's using compute shaders to run code on the GPU, I'm fairly certain it's going to be the same as angrylion but massively parallel.

**Comment by LegendOfDragoon** [comment]
>I'm not sure that it is what we need.

It's not. Better to just learn from Angrylion's source because like you said, it's the most complete and perfectly working implementation.

**Comment by theboy181** [comment]
Killer Instinct Gold (U) (V1.0)
Sky Fix for GLideN64

81019AEA 4100

Note: Works in game, but there are still issues in the menus

**Comment by olivieryuyu** [comment]
> 
> 
> This issue will not gone until somebody will correct all issues with LLE Triangle.

@gonetz 

May be I am mistaken but the sky is not drawn by LLE triangles as for GE 007 and PD.

I cannot see the usage of G_RDPHALF_CONT for instance anywhere.

I see a lot of G_POPMTX used. May be a sign of a core issue of some sort.

Now GlideN64 has some HLE issue with the game. For instance in the demo with Tusk

in HLE GlideN64:

![hle](https://user-images.githubusercontent.com/8987490/49833832-ff035700-fd9a-11e8-97a2-02208f7f8c9f.jpg)

Texture are missing somehow.

With Glide64:

![glide64](https://user-images.githubusercontent.com/8987490/49833841-07f42880-fd9b-11e8-9d7b-acac06031832.jpg)

With GlideN64 LLE:

![lle](https://user-images.githubusercontent.com/8987490/49833888-3114b900-fd9b-11e8-97ae-5821fc646c26.png)

Same than Glide64.

With Angrylion, it is slighty different:

![lle angrylion](https://user-images.githubusercontent.com/8987490/49833907-4a1d6a00-fd9b-11e8-90cf-278ff5981cf5.jpg)

**Comment by olivieryuyu** [comment]
something odd as well is the the 1st word of G_TRI as some data?

**Comment by olivieryuyu** [comment]
ucode seems perfectly standard. Seems a LLE issue.

**Comment by theboy181** [comment]
Killer Instinct V1.0 (U)
Sky Fix

81019AEA 4119

Note: Fixes the sky in GLideN64

I wrote this hack a while back, unsure if it helps you in anyway.

**Comment by gonetz** [comment]
> May be I am mistaken but the sky is not drawn by LLE triangles as for GE 007 and PD.

True. It may sound weird, but plugin's API function ProcessRDPList is called in HLE mode. This function is called for LLE list. RDP_TriShadeTxtr command 0x0E is called from that dlist.
I implemented basic support for LLE display list in Glide64 to emulate sky in KI. GLideN64 has full support for LLE display list.
So yes, it is pure LLE issue.

**Comment by loganmc10** [comment]
I think I know the case where ProcessRDPList would be called in HLE. Would it be helpful if the emulator called the HLE version instead?

**Comment by gonetz** [comment]
If HLE version will contain usual HLE triangle commands, it sure will help.
If it will be call of of LLE triangle via  G_RDPHALF_CONT as in GE 007 and PD, it hardly will help.

**Comment by olivieryuyu** [comment]
So the sky is not part of the usual graphic list? This is why I cannot see them...

There is a separated task for drawing the sky, which could be only by using LLE triangles. 

@loganmc10 

Would be nice if you could find out such a list, though I guess it is directly for LLE triangles.

**Comment by loganmc10** [comment]
@gonetz @olivieryuyu Can you try this build:

https://drive.google.com/file/d/1Cg_61WO_wHrzTYcC5kLAXrw67zKaB3rS/view?usp=sharing

That is a build of mupen64plus that should try to draw the skies using HLE instead of LLE. When I run that, I get this message from GLideN64:
```
[GLideN64]: error - unknown ucode!!!
```

So I'm hoping that either the ucode just needs to get defined with a CRC, or maybe it is a new ucode that can be decoded? I'm not too familiar with that kind of thing

**Comment by olivieryuyu** [comment]
this is totally normal to have such  a message if it is specific ucode. 

Now I can only decode with PJ64. Is it possible to do this for PJ64 with the current PJ64 RSP?

Thanks

**Comment by loganmc10** [comment]
I'm not too familiar with PJ64, but I know what needs to be changed, perhaps @project64 or @LegendOfDragoon could provide a modified build:

https://github.com/project64/project64/blob/56a10af58064f2264e8603ce3de252bd4fd3ef69/Source/Project64-core/N64System/Mips/MemoryVirtualMem.cpp#L1768-L1778

This is where ProcessRDPList (LLE) is called. I commented out that call, and in the ```case 0x04100000``` block, I added a ProcessDList call (HLE)

**Comment by loganmc10** [comment]
Keep in mind that this may be a dead end, it could be that GLideN64 is complaining about ```unknown ucode``` because there is no ucode provided, and this isn't a workable solution

**Comment by Frank-74** [comment]
> This is where ProcessRDPList (LLE) is called. I commented out that call, and in the `case 0x04100000` block, I added a ProcessDList call (HLE)

@olivieryuyu  I made that change and built.  Here it is.
https://www.dropbox.com/s/3hvcfzt2npg284a/Project64PDL.zip?dl=1

**Comment by theboy181** [comment]
@frank to be clear this fixes the issue in PJ? Are there other games that benefit from this?

**Comment by olivieryuyu** [comment]
It does not seem to work

**Comment by Frank-74** [comment]
@theboy181 doesn't fix anything.  Causes no display list, black screen.  I just made the change @loganmc10 suggested.  I think the RSP would need changing as well.

**Comment by theboy181** [comment]
@olivieryuyu It like the background for the sky is stretched out, causing it to render incorrectly.

**Comment by gonetz** [comment]
Fixed
